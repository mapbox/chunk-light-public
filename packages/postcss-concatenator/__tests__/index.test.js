'use strict';

jest.mock('got', () => jest.fn());

const fs = require('fs');
const path = require('path');
const tempy = require('tempy');
const del = require('del');
const fastGlob = require('fast-glob');
const got = require('got');
const PostcssConcatenator = require('..');

const DIR_FIXTURES = path.join(__dirname, 'fixtures');
const fixture = x => path.join(DIR_FIXTURES, x);

const cache = new Map();

afterEach(() => {
  cache.clear();
});

describe('with inline source map', () => {
  let tmp;
  let outputPath;
  beforeAll(() => {
    tmp = tempy.directory();
    outputPath = path.join(tmp, 'output.css');
    return PostcssConcatenator.concatToFile({
      urlCache: cache,
      stylesheets: [
        fixture('a.css'),
        fixture('nested/c.css'),
        fixture('b.css')
      ],
      output: outputPath
    });
  });

  afterAll(() => {
    return del(tmp, { force: true });
  });

  test('output contains each CSS file in order, with an inline source map', () => {
    expect(fs.readFileSync(outputPath, 'utf8')).toMatchSnapshot();
  });

  test('source map not written', () => {
    expect(() => {
      fs.readFileSync(path.join(tmp, 'output.css.map'));
    }).toThrow();
  });

  test('image is copied and hashed', () => {
    const pngFilename = fastGlob.sync([path.join(tmp, 'something_*.png')])[0];
    expect(pngFilename).toBeTruthy();
  });

  test('CSS references hashed image relativized to output directory', () => {
    const pngFilename = fastGlob.sync([path.join(tmp, 'something_*.png')])[0];
    const pngBasename = path.basename(pngFilename);
    expect(fs.readFileSync(outputPath, 'utf8')).toMatch(
      `background-image: url('${pngBasename}')`
    );
  });

  test('nice error when PostCSS fails to parse', () => {
    expect(
      PostcssConcatenator.concatToFile({
        urlCache: cache,
        stylesheets: [fixture('invalid.css')],
        output: path.join(tmp, 'output.css')
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});

describe('with external source map', () => {
  let tmp;
  let outputPath;
  beforeAll(() => {
    tmp = tempy.directory();
    outputPath = path.join(tmp, 'output.css');
    return PostcssConcatenator.concatToFile({
      urlCache: cache,
      stylesheets: [
        fixture('a.css'),
        fixture('nested/c.css'),
        fixture('b.css')
      ],
      output: outputPath,
      sourceMap: 'file'
    });
  });

  afterAll(() => {
    return del(tmp, { force: true });
  });

  test('output contains each CSS file in order, with an external source map', () => {
    expect(fs.readFileSync(outputPath, 'utf8')).toMatchSnapshot();
  });

  test('source map written', () => {
    expect(
      fs.readFileSync(path.join(tmp, 'output.css.map'), 'utf8')
    ).toMatchSnapshot();
  });
});

describe('with a URL', () => {
  let tmp;
  let outputPath;
  beforeAll(() => {
    tmp = tempy.directory();
    outputPath = path.join(tmp, 'output.css');

    got.mockImplementation(getUrl => {
      if (getUrl === 'https://www.foo.com/bar.css') {
        return Promise.resolve({ body: '#bar { color: yellow; }' });
      }
      if (getUrl === 'https://www.foo.com/baz.css') {
        return Promise.resolve({ body: '#baz { color: pink; }' });
      }
      throw new Error(`Unexpected URL ${getUrl}`);
    });

    return PostcssConcatenator.concatToFile({
      urlCache: cache,
      stylesheets: [
        'https://www.foo.com/bar.css',
        fixture('a.css'),
        'https://www.foo.com/baz.css'
      ],
      output: outputPath,
      sourceMap: 'file'
    });
  });

  afterAll(() => {
    return del(tmp, { force: true });
  });

  test('output contains each CSS file in order', () => {
    expect(fs.readFileSync(outputPath, 'utf8')).toMatchSnapshot();
  });

  test('URLs are cached in memory so not fetched again', () => {
    const options = {
      stylesheets: [
        'https://www.foo.com/bar.css',
        fixture('a.css'),
        'https://www.foo.com/baz.css'
      ],
      output: outputPath
    };
    expect(got).toHaveBeenCalledTimes(0);
    return PostcssConcatenator.concatToFile(options)
      .then(() => {
        expect(got).toHaveBeenCalledTimes(2);
        return PostcssConcatenator.concatToFile(options);
      })
      .then(() => {
        expect(got).toHaveBeenCalledTimes(2);
      });
  });
});

test('errors if no stylesheets are provided', () => {
  expect(
    PostcssConcatenator.concatToFile({
      urlCache: cache,
      stylesheets: [],
      output: 'output.css'
    })
  ).rejects.toThrow('No stylesheets');
});
