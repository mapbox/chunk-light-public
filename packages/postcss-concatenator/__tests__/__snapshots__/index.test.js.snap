// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with a URL output contains each CSS file in order 1`] = `
"#bar { color: yellow; }#a {
  color: blue;
}#baz { color: pink; }
/*# sourceMappingURL=output.css.map */"
`;

exports[`with external source map output contains each CSS file in order, with an external source map 1`] = `
"#a {
  color: blue;
}#c {
  color: purple;
  background-image: url('something_2cc5d05.png');
}#b {
  color: red;
}

/*# sourceMappingURL=output.css.map */"
`;

exports[`with external source map source map written 1`] = `"{\\"version\\":3,\\"sources\\":[\\"../../../../../../..<PROJECT_ROOT>/packages/postcss-concatenator/__tests__/fixtures/a.css\\",\\"../../../../../../..<PROJECT_ROOT>/packages/postcss-concatenator/__tests__/fixtures/nested/c.css\\",\\"../../../../../../..<PROJECT_ROOT>/packages/postcss-concatenator/__tests__/fixtures/b.css\\"],\\"names\\":[],\\"mappings\\":\\"AAAA;EACE,YAAY;CACb,ACFD;EACE,cAAc;EACd,+CAAsD;CACvD,ACHD;EACE,WAAW;CACZ\\",\\"file\\":\\"output.css\\",\\"sourcesContent\\":[\\"#a {/n  color: blue;/n}/n\\",\\"#c {/n  color: purple;/n  background-image: url('./nested-again/something.png');/n}/n\\",\\"#b {/n  color: red;/n}/n\\"]}"`;

exports[`with inline source map nice error when PostCSS fails to parse 1`] = `
"PostCSS error: <PROJECT_ROOT>/packages/postcss-concatenator/__tests__/fixtures/invalid.css:2:3: Unknown word
 [90m 1 | [39m[35m#invalid[39m [33m{[39m
[31m[1m>[22m[39m[90m 2 | [39m  foo 7
 [90m   | [39m  [31m[1m^[22m[39m
 [90m 3 | [39m[33m}[39m
 [90m 4 | [39m"
`;

exports[`with inline source map output contains each CSS file in order, with an inline source map 1`] = `
"#a {
  color: blue;
}#c {
  color: purple;
  background-image: url('something_2cc5d05.png');
}#b {
  color: red;
}

/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIi4uLy4uLy4uLy4uLy4uLy4uLy4uL1VzZXJzL2RhdmlkY2xhcmsvZGV2L21hcGJveC91bmRlcnJlYWN0L3BhY2thZ2VzL3Bvc3Rjc3MtY29uY2F0ZW5hdG9yL19fdGVzdHNfXy9maXh0dXJlcy9hLmNzcyIsIi4uLy4uLy4uLy4uLy4uLy4uLy4uL1VzZXJzL2RhdmlkY2xhcmsvZGV2L21hcGJveC91bmRlcnJlYWN0L3BhY2thZ2VzL3Bvc3Rjc3MtY29uY2F0ZW5hdG9yL19fdGVzdHNfXy9maXh0dXJlcy9uZXN0ZWQvYy5jc3MiLCIuLi8uLi8uLi8uLi8uLi8uLi8uLi9Vc2Vycy9kYXZpZGNsYXJrL2Rldi9tYXBib3gvdW5kZXJyZWFjdC9wYWNrYWdlcy9wb3N0Y3NzLWNvbmNhdGVuYXRvci9fX3Rlc3RzX18vZml4dHVyZXMvYi5jc3MiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQUE7RUFDRSxZQUFZO0NBQ2IsQUNGRDtFQUNFLGNBQWM7RUFDZCwrQ0FBc0Q7Q0FDdkQsQUNIRDtFQUNFLFdBQVc7Q0FDWiIsImZpbGUiOiJvdXRwdXQuY3NzIiwic291cmNlc0NvbnRlbnQiOlsiI2Ege1xuICBjb2xvcjogYmx1ZTtcbn1cbiIsIiNjIHtcbiAgY29sb3I6IHB1cnBsZTtcbiAgYmFja2dyb3VuZC1pbWFnZTogdXJsKCcuL25lc3RlZC1hZ2Fpbi9zb21ldGhpbmcucG5nJyk7XG59XG4iLCIjYiB7XG4gIGNvbG9yOiByZWQ7XG59XG4iXX0= */"
`;
